### PR TITLE
[MPS] Fix bidirectional LSTM & small one-direction LSTM fix

### LINKED ISSUE #94754: BiLSTM incorrect on MPS
### 🐛 Describe the bug

Hi, I been testing various NLP models on MPS, and I observe various strange things. It seems that bidirectional LSTM on MPS is not working correctly.

```
import torch
import torch.nn as nn

print("Torch version: ", torch.__version__)
torch.manual_seed(123)
embeddings = torch.rand(3,1,2)  # L, B, E

model = nn.LSTM(input_size=2, hidden_size=2, bidirectional=True)
print(model(embeddings)[0])

model.to("mps")
print(model(embeddings.to("mps"))[0])
```

Results are below. CPU above, MPS below. How can results on CPU be different than on MPS?

```
Torch version:  2.0.0.dev20230205
tensor([[[ 0.0054,  0.2338, -0.1167,  0.2263]],
        [[ 0.0200,  0.3305, -0.0967,  0.2075]],
        [[ 0.0422,  0.3740, -0.0590,  0.1553]]], grad_fn=<CatBackward0>)
tensor([[[0.2325, 0.2447, 0.4149, 0.4394]],
        [[0.3524, 0.3702, 0.3516, 0.3785]],
        [[0.4139, 0.4346, 0.2306, 0.2531]]], device='mps:0',
       grad_fn=<CatBackward0>)
```

### Versions

I’m testing on MacBook Pro with M1PRO, 32GB, 10-core CPU, 16-core GPU.

```
PyTorch version: 2.0.0.dev20230205
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.16 (main, Jan 11 2023, 10:02:19)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0.dev20230205
[pip3] torchaudio==2.0.0.dev20230205
[pip3] torchvision==0.15.0.dev20230205
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  h0a44026_0    pytorch
[conda] mkl                       2021.4.0           hecd8cb5_637  
[conda] mkl-service               2.4.0            py39h9ed2024_0  
[conda] mkl_fft                   1.3.1            py39h4ab4a9b_0  
[conda] mkl_random                1.2.2            py39hb2f4e1b_0  
[conda] numpy                     1.23.5           py39he696674_0  
[conda] numpy-base                1.23.5           py39h9cd3388_0  
[conda] pytorch                   2.0.0.dev20230205         py3.9_0    pytorch-nightly
[conda] torchaudio                2.0.0.dev20230205        py39_cpu    pytorch-nightly
[conda] torchvision               0.15.0.dev20230205        py39_cpu    pytorch-nightly
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by alexdremov**
Fixes #94754

With this PR I hope to finish my breathtaking journey of fixing MPS LSTM.

Here, I enable `bidirectional` on MPS. Also, I've noticed that cache key did not account for all parameters, so there could have been problems with one-directional LSTM when created without bias or dropout and then with one of them.

cc: @kulinseth @jhavukainen

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 1]
```diff
@@ -23,6 +23,85 @@
...
+ * stateTensor, cellStateTensor, recurrentWeight, inputWeight, biasTensor
+ */
+static std::tuple<MPSGraphTensor*, MPSGraphTensor*, MPSGraphTensor*, MPSGraphTensor*, MPSGraphTensor*>
+    getMPSTensorsFromPytorchTensors(MPSGraph* mpsGraph,
+                                MPSGraphTensor* stateTensor, MPSGraphTensor* cellStateTensor,
+                                NSMutableArray<MPSGraphTensor*> *recurrentKernelWeightsList,
+                                NSMutableArray<MPSGraphTensor*> *kernelWeightsList,
+                                NSMutableArray<MPSGraphTensor*> *kernelBiasList,
+                                NSMutableArray<MPSGraphTensor*> *recurrentBiasList,
+                                bool has_biases, bool bidirectional, size_t layer_no) {
```
Comment (AlexRoar) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:38]: This substantial part of code is duplicated in forward and backward, so incapsulated it in function

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 2]
```diff
@@ -65,7 +143,7 @@
     MPSStream* stream = getCurrentMPSStream();
 
     @autoreleasepool {
-      string key = "lstm_" + getTensorsStringKey({input, hx[0], hx[1]}) + getMPSTypeString(input.scalar_type()) + "_num_layers_" + std::to_string(num_layers);
+      string key = "lstm_" + getTensorsStringKey({input, hx[0], hx[1]}) + getMPSTypeString(input.scalar_type()) + "_num_layers_" + std::to_string(num_layers) + "_bidirectional_" + std::to_string(bidirectional) + "_has_biases_" + std::to_string(has_biases) + "_dropout_" + std::to_string(dropout_p) + "_batch_first_" + std::to_string(batch_first);
```
Comment (AlexRoar) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:146]: Here, bug appeared when LSTM was created without biases and then with biases. Added test case

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 3]
```diff
@@ -274,22 +361,19 @@
...
+      if(has_biases) {
+          biases.push_back(params[i*stride + 2]);
+          recurrent_biases.push_back(params[i*stride + 3]);
+      }
     }
 
     struct CachedGraph : public MPSCachedGraph {
       CachedGraph(MPSGraph *graph) : MPSCachedGraph(graph) {}
       std::vector<MPSGraphTensor*> inputTensors_;
-      std::vector<MPSGraphTensor*> outputTensors_;
```
Comment (AlexRoar) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:292]: Turns out, I left it unused in the previous PR

File: test/test_mps.py [hunk 4]
```diff
@@ -9054,7 +9124,7 @@ def get_results(device, inp, hx, cx):
                 inp, hx, cx = inp.to(device), hx.to(device), cx.to(device)
 
                 output, _ = rnn(inp, (hx, cx))
-                f = output.sum()
+                f = output.sum() + hx.sum() + cx.sum()
```
Comment (AlexRoar) [inline test/test_mps.py:9127]: Added `hx` and `cx` to function to test how gradients flow through them

File: test/test_mps.py [hunk 5]
```diff
@@ -9054,7 +9124,7 @@ def get_results(device, inp, hx, cx):
                 inp, hx, cx = inp.to(device), hx.to(device), cx.to(device)
 
                 output, _ = rnn(inp, (hx, cx))
-                f = output.sum()
+                f = 3 * output.sum() + (hx * cx).sum()
```
Comment (AlexRoar) [inline test/test_mps.py:9127]: Added hx and cx to function to test how gradients flow through them

File: aten/src/ATen/native/mps/operations/RnnOps.mm (hunk 2 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/RnnOps.mm:146]: Good catch

**Review by jhavukainen — APPROVED** [review]
All looks good to me, well done. Nice catch on the incorrect caching behavior with the biases.

**Comment by kulinseth** [comment]
Looks great.

**Review by kulinseth — APPROVED** [review]

**Comment by AlexRoar** [comment]
@kulinseth merge failed(

**Comment by AlexRoar** [comment]
Just rebased

**Comment by kulinseth** [comment]
@albanD , @malfet , @soulitzer , can one of you please take a quick look at this ? MPS changes look good. Need approval from Core maintainers on this.

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 6]
```diff
@@ -556,25 +737,26 @@
...
-
             gradRecWeightsPlaceholder = Placeholder([gradRecWeightsArray objectAtIndex: i], grad_rec_weights);
             gradWeightsPlaceholder = Placeholder([gradWeightsArray objectAtIndex: i], grad_weights);
-            gradBiasPlaceholder = Placeholder([gradBiasArray objectAtIndex: i], grad_bias);
 
-            [results setObject:gradBiasPlaceholder.getMPSGraphTensorData() forKey:gradBiasPlaceholder.getMPSGraphTensor()];
             [results setObject:gradRecWeightsPlaceholder.getMPSGraphTensorData() forKey:gradRecWeightsPlaceholder.getMPSGraphTensor()];
             [results setObject:gradWeightsPlaceholder.getMPSGraphTensorData() forKey:gradWeightsPlaceholder.getMPSGraphTensor()];
+
+            if(has_biases) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:753]: Nit
```suggestion
            if (has_biases) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 7]
```diff
@@ -556,25 +737,26 @@
         Placeholder gradRecWeightsPlaceholder, gradWeightsPlaceholder, gradBiasPlaceholder;
 
         std::vector<Tensor> weights;
-        for (int i = 0; i < num_layers; i++) {
+        for (int i = 0; i < num_layers * (bidirectional ? 2 : 1); i++) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:740]:
```suggestion
        for (const auto  i : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 8]
```diff
@@ -514,18 +698,15 @@
...
         NSMutableArray<MPSGraphTensor*> *recurrentBiasList = cachedGraph->recurrentBiasList_;
-        Placeholder kernelWeight;
-        Placeholder recurrentKernelWeight;
-        Placeholder bias;
-        Placeholder recurrentBias;
-        for (size_t i = 0; i < num_layers; i+=1) {
-            kernelWeight = Placeholder([kernelWeightsList objectAtIndex:i], kernel_weights[i]);
-            recurrentKernelWeight = Placeholder([recurrentKernelWeightsList objectAtIndex:i], recurrent_kernel_weights[i]);
+
+        for (size_t i = 0; i < num_layers * (bidirectional ? 2 : 1); i+=1) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:702]:
```suggestion
        for (const auto I : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 9]
```diff
@@ -323,7 +407,7 @@
                     NSMutableArray<MPSGraphTensor*> *kernelBiasList = [[NSMutableArray alloc] initWithCapacity:params.size()];
                     NSMutableArray<MPSGraphTensor*> *recurrentBiasList = [[NSMutableArray alloc] initWithCapacity:params.size()];
 
-                    for (size_t i = 0; i < num_layers; i += 1) {
+                    for (size_t i = 0; i < num_layers * (bidirectional ? 2 : 1); i += 1) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:410]:
```suggestion
                    for (const auto i : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 10]
```diff
@@ -38,15 +117,14 @@
     std::vector<Tensor> recurrent_kernel_weights;
     std::vector<Tensor> biases;
     std::vector<Tensor> recurrent_biases;
-    for (size_t i = 0; i < num_layers; i+=1) {
+    for (size_t i = 0; i < num_layers * (bidirectional ? 2 : 1); i += 1) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:120]:
```suggestion
    for (const auto i : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 11]
```diff
@@ -274,22 +361,19 @@
...
-        if(has_biases) {
-            kernel_weights.push_back(params[i*4]);
-            recurrent_kernel_weights.push_back(params[i*4+1]);
-            biases.push_back(params[i*4+2]);
-            recurrent_biases.push_back(params[i*4+3]);
-        } else {
-            kernel_weights.push_back(params[i*2]);
-            recurrent_kernel_weights.push_back(params[i*2+1]);
-        }
+    for (size_t i = 0; i < num_layers * (bidirectional ? 2 : 1); i += 1) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:364]:
```suggestion
    for (const auto i : c10::irange(num_layers * (bidirectional ? 2 : 1)) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 12]
```diff
@@ -556,25 +737,26 @@
...
-            gradBiasPlaceholder = Placeholder([gradBiasArray objectAtIndex: i], grad_bias);
 
-            [results setObject:gradBiasPlaceholder.getMPSGraphTensorData() forKey:gradBiasPlaceholder.getMPSGraphTensor()];
             [results setObject:gradRecWeightsPlaceholder.getMPSGraphTensorData() forKey:gradRecWeightsPlaceholder.getMPSGraphTensor()];
             [results setObject:gradWeightsPlaceholder.getMPSGraphTensorData() forKey:gradWeightsPlaceholder.getMPSGraphTensor()];
+
+            if(has_biases) {
+                Tensor grad_bias = at::empty((kernel_weights[i].size(0)), kernel_weights[i].options());
+                weights.push_back(grad_bias);
+                weights.push_back(grad_bias);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:756]: Not super familiar with the code, would be good to leave a comment why `grad_bias` should be pushed twice.

**Review by malfet — APPROVED** [review]
LGTM, few general remarks:
- Please use `for(const auto i : c10::irange(limit))` to iterate over list of elements
- `num_layer * (bidirectional ? 2 :1 )` is used frequently in the code, perhaps a helper variable , say `total_num_layer` can be defined at the start of the routine and then user throughout the codebase?

File: aten/src/ATen/native/mps/operations/RnnOps.mm (hunk 12 above)
Comment (AlexRoar) [reply aten/src/ATen/native/mps/operations/RnnOps.mm:756]: Sure. I've added several comments near complicated pieces. Will extend them in commit with your fixes

Comment (AlexRoar) [reply aten/src/ATen/native/mps/operations/RnnOps.mm:768]: Added explanation
